src/Bounds.cpp

```diff
@@ -99,61 +99,59 @@ class Bounds : public IRVisitor {
         FuncValueBounds::const_iterator iter = func_bounds.find(key);
 
         if (iter != func_bounds.end()) {
-            if (iter->second.min.defined()) {
-                min = iter->second.min;
+            if (iter->second.has_lower_bound()) {
+                interval.min = iter->second.min;
             }
-            if (iter->second.max.defined()) {
-                max = iter->second.max;
+            if (iter->second.has_upper_bound()) {
+                interval.max = iter->second.max;
             }
         }
     }
 
     void bounds_of_type(Type t) {
         t = t.element_of();
         if (t.is_uint() && t.bits() <= 16) {
-            max = cast(t, (1 << t.bits()) - 1);
-            min = cast(t, 0);
+            interval = Interval(cast(t, 0),
+                                cast(t, (1 << t.bits()) - 1));
+
         } else if (t.is_int() && t.bits() <= 16) {
-            max = cast(t, (1 << (t.bits()-1)) - 1);
-            min = cast(t, -(1 << (t.bits()-1)));
+            interval = Interval(cast(t, -(1 << (t.bits()-1))),
+                                cast(t, (1 << (t.bits()-1)) - 1));
+
         } else {
-            max = Expr();
-            min = Expr();
+            interval = Interval::everything();
         }
     }
 
     using IRVisitor::visit;
 
     void visit(const IntImm *op) {
-        min = op;
-        max = op;
+        interval = Interval::single_point(op);
     }
 
     void visit(const UIntImm *op) {
-        min = op;
-        max = op;
+        interval = Interval::single_point(op);
     }
 
     void visit(const FloatImm *op) {
-        min = op;
-        max = op;
+        interval = Interval::single_point(op);
     }
 
     void visit(const Cast *op) {
 
         op->value.accept(this);
-        Expr min_a = min, max_a = max;
+        Interval a = interval;
 
-        if (min_a.same_as(op->value) && max_a.same_as(op->value)) {
-            min = max = op;
+        if (a.is_single_point(op->value)) {
+            interval = Interval(op, op);
```

Comment:
What guides the choice to use "Interval(op, op)" instead of "Interval::single_point(op)"?